docs/incubator-proposal.md

```diff
@@ -0,0 +1,194 @@
+these certificates, the end goal has largely been the same. To provide a signed
+keypair in a Secret resource to then be consumed, either by an Ingress
+controller or some other workload in-cluster, and then keep that keypair up to
+date.
+
+Looking beyond the capabilities of all of these projects, it'd also be
+beneficial for users to be able to utilise alternative issuer backends, other
+than ACME. This could include [Hashicorp Vault](https://vaultproject.io), a
+plain CA (backed by Kubernetes secret resources), or any other cloud based
+service.
+
+It is therefore proposed that a new project, `cert-manager`, is created that
+aims to generalise the issuance and renewal of certificates from an arbitrary
+certificate source. This project should be designed in a similar manner to
+other Kubernetes controllers (eg. kube-controller-manager, service-catalog et
+al.).
+
+## Definitions
+
+* `Issuer` - an Issuer is a generic backend that issues certificates. cert-manager
+will contain logic on how to process each issuer, and multiple Issuer resources
+utilising the same issuer implementation may exist (e.g. to allow issuing
+certificates from both letsencrypt production & staging). An example manifest for
+an Issuer resource:
+
+```yaml
+kind: Issuer
+metadata:
+  name: letsencrypt-staging
+spec:
+  acme:
+    # The ACME server URL
+    server: https://acme-staging.api.letsencrypt.org/directory
+    # Email address used for ACME registration
+    email: user@example.com
+    # Name of a secret used to store the ACME account private key
+    privateKey: letsncrypt-staging
+    # ACME dns-01 provider configurations
+    dns-01:
+      # Here we define a list of DNS-01 providers that can solve DNS challenges
+      providers:
+      # We define a provider named 'clouddns', with configuration for the
+      # clouddns challenge provider.
+      - name: clouddns
+        clouddns:
+          # A secretKeyRef to a the google cloud json service account
+          serviceAccount:
+            name: clouddns-service-account
+            key: service-account.json
+          # The project in which to update the DNS zone
+          project: gcloud-project
+      # We define a provider named 'cloudflare', with configuration for the
+      # cloudflare challenge provider.
+      - name: cloudflare
+        cloudflare:
+          # A secretKeyRef to a the cloudflare api key
+          apiKey:
+            name: cloudflare-config
+            key: api-key
+          # The cloudflare user account email
+          email: cloudflare-user@example.com
+```
+
+* `Certificate` - a Certificate resource details a Certificate keypair to
+manage with cert-manager. It contains details such as the hostnames to be listed
+on the certificate, as well as details of which issuer to issuer certificates with,
+and any additional configuration required for the selected issuer. An example
+manifest for a Certificate resource:
```

Comment:
Is the expectation here to use a genuine bundled certificate format (i.e PKCS#12) or to provide certificates and private keys?

In an ideal world, the entities using certificates generate the private keys and CSR's themselves, providing the CSR to the certificate manager service

Reply:
So at the moment we store certificates and private keys in PEM format within secret resources in the target API server. This is because ingress controllers use TLS 'typed' Secret resources (ie. secrets with the `tls.key` and `tls.crt` fields set).

Regarding private keys - right now it is possible for a user to provide their own private key. They'll just need to ensure it exists in the Secret resource before creating the Certificate (if cert-manager see's that a private key does not exist, it automatically generates one right now)

Regarding CSRs - so this is something I am interested in working out. #13 tracks this. It makes sense to me to expose more of the CSR surface, and the `certificates.k8s.io` api group used for kubelet TLS bootstrapping actually does just use a base64 encoded CSR. There's a tradeoff here between UX and functionality and I'm not sure exactly what the sweet spot is. Perhaps we split out a `Certificate` into two resources - one that is a plain old CSR (base64 encoded), the other is what we see today, which essentially provides a thin layer over CSRs (so cert-manager would convert a Certificate resource, with nice and abstracted fields into a CSR). This is discussion for #13 though!